tools/draincluster/drain/drain.go

```diff
@@ -0,0 +1,244 @@
+	}
+
+	if err := h.fetchClusterResourcePlacementToEvict(ctx); err != nil {
+		return false, err
+	}
+
+	if len(h.ClusterResourcePlacementResourcesMap) == 0 {
+		log.Printf("There are currently no resources propagated to %s from fleet using ClusterResourcePlacement resources", h.ClusterName)
+		return true, nil
+	}
+
+	// create eviction objects for all <crpName, targetCluster>.
+	for crpName := range h.ClusterResourcePlacementResourcesMap {
+		evictionName := fmt.Sprintf(drainEvictionNameFormat, crpName, h.ClusterName)
+
+		if err := removeExistingEviction(ctx, h.HubClient, evictionName); err != nil {
+			return false, fmt.Errorf("failed to remove existing eviction for CRP %s", crpName)
+		}
+
+		err := retry.OnError(retry.DefaultBackoff, func(err error) bool {
+			return k8errors.IsAlreadyExists(err)
+		}, func() error {
+			eviction := placementv1beta1.ClusterResourcePlacementEviction{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: evictionName,
+				},
+				Spec: placementv1beta1.PlacementEvictionSpec{
+					PlacementName: crpName,
+					ClusterName:   h.ClusterName,
+				},
+			}
+			return h.HubClient.Create(ctx, &eviction)
+		})
+
+		if err != nil {
+			return false, fmt.Errorf("failed to create eviction for CRP %s: %w", crpName, err)
+		}
+	}
+
+	// wait until all evictions reach a terminal state.
+	for crpName := range h.ClusterResourcePlacementResourcesMap {
+		err := wait.ExponentialBackoffWithContext(ctx, retry.DefaultBackoff, func(ctx context.Context) (bool, error) {
+			evictionName := fmt.Sprintf(drainEvictionNameFormat, crpName, h.ClusterName)
+			eviction := placementv1beta1.ClusterResourcePlacementEviction{}
+			if err := h.HubClient.Get(ctx, types.NamespacedName{Name: evictionName}, &eviction); err != nil {
+				return false, fmt.Errorf("failed to get eviction %s: %w", evictionName, err)
+			}
+			return evictionutils.IsEvictionInTerminalState(&eviction), nil
+		})
+
+		if err != nil {
+			return false, fmt.Errorf("failed to wait for evictions to reach terminal state: %w", err)
+		}
+	}
+
+	isDrainSuccessful := true
+	for crpName := range h.ClusterResourcePlacementResourcesMap {
+		evictionName := fmt.Sprintf(drainEvictionNameFormat, crpName, h.ClusterName)
+		eviction := placementv1beta1.ClusterResourcePlacementEviction{}
+		if err := h.HubClient.Get(ctx, types.NamespacedName{Name: evictionName}, &eviction); err != nil {
+			return false, fmt.Errorf("failed to get eviction %s: %w", evictionName, err)
+		}
+		validCondition := eviction.GetCondition(string(placementv1beta1.PlacementEvictionConditionTypeValid))
+		if validCondition != nil && validCondition.Status == metav1.ConditionFalse {
+			// check to see if CRP is missing or CRP is being deleted or CRB is missing.
+			if validCondition.Reason == condition.EvictionInvalidMissingCRPMessage ||
+				validCondition.Reason == condition.EvictionInvalidDeletingCRPMessage ||
+				validCondition.Reason == condition.EvictionInvalidMissingCRBMessage {
+				continue
```

Comment:
Hi Arvind! If I understand it correctly, in these situations the drain is considered successfully but no output is given?

Reply:
Hi Arvind! There might be some misunderstanding here; I was trying to suggest that we probably should add a log here mentioning that the CRP is already gone, just for better visibility.